### PR TITLE
tests: add web platform test runner

### LINKED ISSUE #6229: discussion: Run web platform tests
[WPT](https://github.com/web-platform-tests/wpt) is an extensive suite of tests that tests nearly all web platform features like `fetch`, `console`, and streams. It would be great if we could run the for us relevant tests during CI. Chrome and Firefox run WPT during testing too.

### PR DESCRIPTION
**Pull request by lucacasonato**
Closes #6229

cc @kitsonk I only enabled a minimal set of tests for `streams`, because most of the rest does not work. We can start enabling them one by one in follow ups.

Files are all concatenated together, because WPT tests are meant to work with scripts, not modules. ~Once we support external source mapping, I will add source mapping for WPT tests.~ I have added source mapping. I have checked that it works with #8995.

### CODE REVIEW

File: cli/tests/wpt_testharnessconsolereporter.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    console.log(
+      `\n${result.name}\n${result.message}\n${result.stack}`,
+    );
+  }
+
+  if (failed.length > 0) {
+    console.log(`\nfailures:\n`);
+  }
+  for (const result of failed) {
+    console.log("        " + result.name);
```
Comment (kitsonk) [inline cli/tests/wpt_testharnessconsolereporter.js:104]:
```suggestion
    console.log(`        ${result.name}`);
```

File: cli/tests/wpt_testharnessconsolereporter.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+export function yellow(str) {
+  return run(str, code([33], 39));
+}
+
+const testResults = [];
+const testsExpectFail = JSON.parse(Deno.args[0]);
+
+window.add_result_callback((test) => {
```
Comment (kitsonk) [inline cli/tests/wpt_testharnessconsolereporter.js:33]:
```suggestion
window.add_result_callback(({ message, name, stack, status }) => {
```

And then you can just use them destructured below...

File: cli/tests/wpt_testharnessconsolereporter.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  console.log(simpleMessage);
+
+  testResults.push({
+    name: test.name,
+    passed: test.status === 0,
+    expectFail,
+    message: test.message,
+    stack: test.stack,
+  });
+});
```
Comment (kitsonk) [inline cli/tests/wpt_testharnessconsolereporter.js:76]: If you do the above...

```suggestion
  testResults.push({
    name,
    passed: status === 0,
    expectFail,
    message,
    stack,
  });
});
```

**Review by kitsonk — COMMENTED** [review]
A couple thoughts/nits

File: .gitmodules [hunk 4]
```diff
@@ -5,3 +5,6 @@
 [submodule "std/wasi/testdata"]
 	path = std/wasi/testdata
 	url = https://github.com/khronosproject/wasi-test-suite.git
+[submodule "cli/tests/wpt"]
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply .gitmodules:8]: Maybe into `test_util`? We wouldn't need a new top level directory then.

[the remark this answers is not in the document]
Comment (kitsonk) [reply .gitmodules:8]: 👍

**Review by kitsonk — APPROVED** [review]
Awesome! LGTM.

**Comment by axetroy** [comment]
This is really awesome!

If you can output the test results to a JSON file and render them in `https://deno.land`, that would be great

**Comment by lucacasonato** [comment]
> If you can output the test results to a JSON file and render them in `https://deno.land`, that would be great

Sounds like a good idea. Will investigate.

**Pull request merged**